tests/dat/actions/concurrent.js

```diff
@@ -17,6 +19,10 @@ function main(args) {
             setTimeout(function() {
                 checkRequests(args, resolve, reject);
             }, interval);
+        }).finally(() => {
+            // Before leaving the container, decrement the counter again. But wait twice the interval, before decrementing it. Otherwise, the
+            // other requests might not realise, that all requests were inside the container.
+            setTimeout(() => counter--, 2 * interval);
```

Comment:
when is finally executed - ie is there a guarantee the container isn't paused first?

Reply:
Finally is executed after the promise, regardless of the promises fade. It is the equivalent of scalas `Future.andThen { ... }`.

Reply:
I think @rabbah's comment is towards: Is it guaranteed to be executed before the next `.then` block? (Scala's `andThen` forces that ordering).

Reply:
Is there any reason not to:
- remove the finally (and the `setTimeout` within)
- add `counter=0;` to immediately before `resolve(result)` (within `checkRequests`)
?
This way there is no timing issue (potentially racing pause grace timeout), and the counter always gets reset unless a reject occurs.

Reply:
@markusthoemmes @rabbah 
If I understand the following documentation of http://bluebirdjs.com/docs/api/finally.html correctly, I would say, that the finally will always be executed after the next then:
```
If the handler function passed to .finally returns a promise, the promise returned by .finally will not be settled until the promise returned by the handler is settled. If the handler fulfills its promise, the returned promise will be fulfilled or rejected with the original value. If the handler rejects its promise, the returned promise will be rejected with the handler's value.
```
As finally returns a new promise, which is required for the next `.then`-block, I should be executed first.

@tysonnorris 
I don't know, if I get your proposal completely, but this action is also used for other tests, which expect, that there are concurrent requests within the same container.
By removing the `setTimeout` before decrementing, only one of these requests has the chance to realise, that all expected requests are inside the container now. By setting the counter to 0, this would not be the correct behaviour for tests, that expect several requests inside this container. And adapting the counter only in success case would also be the wrong behaviour.

Reply:
@cbickel How about:
- add a new variable to track pending requests count
- decrement that before resolve
- reset counter when pending reaches 0
```        pending--;
        if (pending == 0){
            counter=0;
        } 
```
Just trying to make it even less sensitive to timing configs.

> And adapting the counter only in success case would also be the wrong behaviour.

Is it? In case of rejections, won't the container not be used as warm anymore (should be destroyed instead of reused)?

Reply:
I see the following problems with this implementation:
- the counter doesn't represent the amount of requests in the container anymore.
- as the action is only failed with an application-error, the container will not be removed and will potentially be reused. So we need to duplicate the code of decrementing the counter.

Reply:
Not sure I follow, this seems to work testing locally.
> the counter doesn't represent the amount of requests in the container anymore. 
How so? we only reset counter at time of resolve, and resolve only if the counter is expected value (n in case of n concurrent supported activations)
 >  as the action is only failed with an application-error, the container will not be removed
I see - I get the error behavior wrong all the time :(...  so I guess regardless of any reject happening, the test where reject occurs will fail, and within that test, at least the rejected activation will fail, but additionally another may fail (due to container reuse), which seems ok within the same action+test? (the tests all use uniquely named actions, so there should not be reuse between tests).
WDYT?